src/main/java/me/magnet/consultant/RoutingStrategies.java

```diff
@@ -0,0 +1,110 @@
+package me.magnet.consultant;
+
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+public class RoutingStrategies {
+
+	public static final RoutingStrategy RANDOMIZED_WEIGHTED_DISTANCE = new RoutingStrategy() {
+
+		private final Random random = new Random();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> instances = serviceLocator.listInstances(serviceName);
+			if (instances.size() <= 1) {
+				// No routing to do.
+				return instances;
+			}
+
+			// Pick a random number between 0-1.
+			double chance = random.nextDouble();
+
+			/*
+			 * Pick the index of the first instance to try based on this chance.
+			 * If the chance was < 0.5 then use the instance on index 0.
+			 * If the chance was >= 0.5 and < 0.75 then use the instance on index 1.
+			 * If the chance was >= 0.75 and < 0.875 then use the instance on index 2.
+			 * If the chance was >= 0.875 and < 0.9375 then use the instance on index 3.
+			 * Etc...
+			 */
+			int index = 0;
+			double threshold = 0.5;
+			while (threshold < chance && index < instances.size() - 1) {
+				index++;
+				threshold = Math.pow(0.5, index + 1);
+			}
+
+			return splice(instances, index);
+		}
+	};
+
+	public static final RoutingStrategy ROUND_ROBIN = new RoutingStrategy() {
+
+		private final Multimap<String, ServiceInstance> lastRequested = HashMultimap.create();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> nextAvailable = serviceLocator.listInstances(serviceName);
+			Set<ServiceInstance> lastRequests = Sets.newHashSet(lastRequested.get(serviceName));
+
+			// If all instances have been hit at least once, reset the counter for this service.
+			if (lastRequests.containsAll(nextAvailable)) {
+				lastRequested.removeAll(serviceName);
+			}
+
+			// List the instances to hit in order of not hit to already hit (conserving network distance).
+			List<ServiceInstance> nextUp = nextAvailable.stream()
+					.filter(next -> !lastRequests.contains(next))
+					.collect(Collectors.toList());
+			nextAvailable.stream()
+					.filter(lastRequests::contains)
+					.forEachOrdered(nextUp::add);
+
+			// Register that nextUp[0] instance is going to get hit first.
+			if (!nextUp.isEmpty()) {
+				lastRequested.put(serviceName, nextUp.get(0));
+			}
+
+			return nextUp;
+		}
+	};
```

Comment:
1. There's a couple of considerations that made it the way it is. First of all is that we simply don't know which item in the list is going to be used. Convention is to use the first one, and only use the next item if the previous one failed. In that regard we cannot record failures, and count additional requests against the fallback. We could return a custom `Iterator` with which we could record this, but this is easily subverted by doing `Lists.newArrayList(iterator)`, but this involves changing the API of Consultant. Returning one item is already done by the `locate()` method in `Consultant`, but that doesn't allow for nice failovers (or request racing) because it might easily return the same instance twice (we need a bit of state for that). Lastly (and this is an issue we still need to tackle), if no instances are available in the local datacenter, we don't go looking at other nearby datacenters. Reason is because it's kind of a waste to fetch all datacenters (ordered in distance), and then ask each datacenter for their instances (also ordered in distance). The `Iterator`-style pattern would be the best match since it could expand the search beyond the local datacenter when no services are available locally without incurring penalties all the time.

2. The current implementation only guarantees that every service will be hit equally often, it doesn't make any guarantees about ordering. Ordering is implicitly on network distance, which typically doesn't change that much. A follow-up change also deals with services becoming unavailable, btw.